plugins/ccgx/fu-ccgx-dmc-device.c

```diff
@@ -0,0 +1,676 @@
+
+		if (manual_replug)
+			trigger_code = DMC_TRIGGER_CODE_UPDATE_ON_DISCONNECT;
+
+		if (!fu_ccgx_dmc_device_send_download_trigger (self,
+							       trigger_code,
+							       error)) {
+			g_prefix_error (error, "download trigger error: ");
+			return FALSE;
+		}
+	} else if (self->update_model == DMC_UPDATE_MODEL_PENDING_RESET) {
+		if (!fu_ccgx_dmc_device_send_sort_reset (self,
+							 manual_replug,
+							 error)) {
+			g_prefix_error (error, "soft reset error: ");
+			return FALSE;
+		}
+	} else {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid update model = %u",
+			     self->update_model);
+		return FALSE;
+	}
+
+	if (manual_replug)
+		return TRUE;
+
+	fu_device_set_remove_delay (FU_DEVICE (self), self->remove_delay*1000); /* over a minute to flash offline */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_dmc_device_setup (FuDevice *device, GError **error)
+{
+	FuCcgxDmcDevice *self = FU_CCGX_DMC_DEVICE (device);
+	DmcDockStatus dock_status = {0};
+	DmcDockIdentity dock_id = {0};
+	guint32 version_raw = 0;
+	g_autofree gchar *version = NULL;
+	g_autofree gchar *instance_id = NULL;
+
+	/* get dock identity */
+	if (!fu_ccgx_dmc_device_get_dock_id (self, &dock_id, error))
+		return FALSE;
+
+	/* store dock identity */
+	if (!fu_memcpy_safe ((guint8 *)&self->dock_id, sizeof(DmcDockIdentity), 0x0, /* dst */
+		     (guint8*)&dock_id, sizeof(DmcDockIdentity), 0, sizeof(DmcDockIdentity), error)) /* src */
+		return FALSE;
+
+	/* get dock status */
+	if (!fu_ccgx_dmc_device_get_dock_status (self, &dock_status, error))
+		return FALSE;
+
+	/* set instance id */
+	instance_id = g_strdup_printf ("USB\\VID_%04X&PID_%04X",
+					fu_usb_device_get_vid (FU_USB_DEVICE (self)),
+					fu_usb_device_get_pid (FU_USB_DEVICE (self)));
+
+	fu_device_add_instance_id (FU_DEVICE (self), instance_id);
+
+	/* set composite version */
+	version_raw = dock_status.composite_version;
+	version = fu_common_version_from_uint32 (version_raw, FWUPD_VERSION_FORMAT_QUAD);
+	fu_device_set_version (FU_DEVICE (self), version);
+	fu_device_set_version_raw (FU_DEVICE (self), version_raw);
+	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_dmc_device_set_quirk_kv (FuDevice *device,
+				 const gchar *key,
+				 const gchar *value,
+				 GError **error)
+{
+	FuCcgxDmcDevice *self = FU_CCGX_DMC_DEVICE (device);
+
+	if (g_strcmp0 (key, "ImageKind") == 0) {
+		self->fw_image_type = fu_ccgx_fw_image_type_from_string (value);
+		if (self->fw_image_type != FW_IMAGE_TYPE_UNKNOWN)
+			return TRUE;
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid ImageKind");
+		return FALSE;
+	}
+	if (g_strcmp0 (key, "RemoveDelay") == 0) {
```

Comment:
@hughsie I feel like this is the sort of thing that we want in proper daemon quirks, not this custom quirk and custom code that ferries the attribute around only to call `fu_device_set_remove_delay`

Reply:
Added to master, good idea.